### PR TITLE
[READY] Billable hours overview

### PR DESCRIPTION
**Pull request by Marthyn**
- [x] Bill whole project
- [x] Bill single entry
- [x] Filter on client, project and dates
- [x] Filter on billed/not billed
- [x] Mobile
- [x] Caching
- [x] specs
- [x] Scrolling/Sorting

screen shot 2015-01-02 at 16 36 42
screen shot 2015-01-02 at 16 36 52
screen shot 2015-01-02 at 16 37 35

### CODE REVIEW

File: app/controllers/billables_controller.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class BillablesController < ApplicationController
+  def index
+    @projects = Project.all.billable
```
Comment (jurre) [inline app/controllers/billables_controller.rb]: You can (and should) actually call the billable scope directly on Project, so: `Project.billable` :)

File: db/schema.rb [hunk 2]
```diff
@@ -86,6 +86,7 @@
     t.datetime "created_at"
     t.datetime "updated_at"
     t.string   "description"
+    t.boolean  "billed"
```
Comment (jurre) [inline db/schema.rb]: We should probably add an index on this (we can measure this first but just so we don't forget)

File: app/models/entry.rb [hunk 3]
```diff
@@ -20,6 +21,7 @@ class Entry < ActiveRecord::Base
 
   belongs_to :project, touch: true
   belongs_to :category
+  has_one :client, through: :project
```
Comment (jurre) [inline app/models/entry.rb]: Shouldn't this be a `belongs_to` association?

File: app/controllers/billables_controller.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
+class BillablesController < ApplicationController
+
+  def index
+    resource
+    filter_entries if params[:filters]
+    @projects = Project.all
+    @clients = Client.all
```
Comment (jurre) [inline app/controllers/billables_controller.rb]: I'd like to extract a form object for this :)

File: app/controllers/billables_controller.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    filter_params(params).each do |filter, value|
+      @entries = resource.public_send(filter, value) if value.present?
+    end
+  end
+
+  def resource
+    @entries ||= Entry.billable
+  end
+
+  def filter_params(params)
```
Comment (jurre) [inline app/controllers/billables_controller.rb]: I don't think we need this argument right?

**Comment by jurre** [comment]
Awesome-o!

File: app/models/entry.rb (hunk 3 above)
Comment (Marthyn) [inline app/models/entry.rb]: A belongs to through is not possible. But maybe we should add client is on entry level

Comment (jurre) [inline app/models/entry.rb]: No that would get too confusing, this is fine ^^

**Comment by dkhgh** [comment]
Looking good :)

**Comment by Marthyn** [comment]
screen shot 2015-01-05 at 11 29 28
screen shot 2015-01-05 at 11 29 35
screen shot 2015-01-05 at 11 31 03
screen shot 2015-01-05 at 11 31 12
screen shot 2015-01-05 at 11 31 20


Need some mobile filtering on table headers

**Comment by Marthyn** [comment]
screen shot 2015-01-06 at 13 57 59
It now shows how many entries you've marked

screen shot 2015-01-06 at 14 27 29
the filters/bill button scrolls down the page

screen shot 2015-01-06 at 14 27 17
and on mobile after every project there's a bill button

**Comment by Marthyn** [comment]
screen shot 2015-01-06 at 14 31 21
screen shot 2015-01-06 at 14 32 11
screen shot 2015-01-06 at 14 32 20

**Comment by Marthyn** [comment]
rebased

**Comment by dkhgh** [comment]
BTW, I don't see a "select all entries" option on the screenshots, or am I missing something?

**Comment by dkhgh** [comment]
:dancer: nothing to see here

**Comment by Marthyn** [comment]
Per project you can, not for all entries

File: app/views/billables/_entries.html.haml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+%table.entries
+  %th
+    %tr
+      %td Billed
+      %td Date
+      %td Category
+      %td Hours
+      %td User
```
Comment (fatboypunk) [inline app/views/billables/_entries.html.haml]: Should be I18n right?

Comment (Marthyn) [inline app/views/billables/_entries.html.haml]: oh definitely, forgot to look at this again

File: app/views/billables/index.html.haml [hunk 7]
```diff
@@ -0,0 +1,16 @@
+.billables
+  .sidebar
+    %h1 Filters
+    .container
+      = render "filters"
+  .outer
+    %h1 Billable Entries
```
Comment (fatboypunk) [inline app/views/billables/index.html.haml]: Here as well btw

**Comment by Marthyn** [comment]
screen shot 2015-01-07 at 15 22 34

**Comment by Marthyn** [comment]
@jurre refactored billable list

**Comment by Marthyn** [comment]
just need some css help tomorrow then it can be merged

**Comment by Marthyn** [comment]
Rebase done :dancers:

File: app/models/entry_query.rb [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+  private
+
+  attr_reader :entries, :params
+
+  def filter_params
+    params ? params.slice(:client_id, :project_id, :billed, :to_date, :from_date).reject { |_, value| value.nil? } : []
+  end
+
+  def present(value)
```
Comment (jurre) [inline app/models/entry_query.rb]: `def present?(value)`

**Comment by Marthyn** [comment]
screen shot 2015-01-09 at 15 11 15
screen shot 2015-01-09 at 15 11 43

**Comment by Marthyn** [comment]
This one can finally be merged :dancer: :dancers:

**Comment by Marthyn** [comment]
YES

**Comment by jurre** [comment]
Super awesome @Marthyn!

**Pull request merged**